File: ide-features-trainer/src/training/commands/Command.java
package training.commands;

import com.intellij.openapi.application.ApplicationManager;
import training.ui.Message;
import training.learn.Lesson;
import training.learn.LessonManager;
import training.util.XmlUtil;

/**
 * Created by karashevich on 30/01/15.
 */
public abstract class Command {


    private CommandType commandType;

        public enum CommandType {START, TEXT, TRY, TRYBLOCK, ACTION, REPLAY, NOCOMMAND, MOVECARET, TYPETEXT, COPYTEXT, TRAVERSECARET, MOUSEBLOCK, MOUSEUNBLOCK, WAIT, CARETBLOCK, CARETUNBLOCK, SHOWLINENUMBER, EXPANDALLBLOCKS, WIN, TEST, SETSELECTION}

    public Command(CommandType commandType) {
        this.commandType = commandType;
    }

    public CommandType getCommandType() {
        return commandType;
    }

    void updateDescription(String s, Lesson lesson){
        LessonManager lessonManager = LessonManager.getInstance(lesson);
        if (lessonManager != null) lessonManager.addMessage(s);
    }

    void updateHTMLDescription(String htmlText, Lesson lesson){
//        if (element.getAttribute("description") != null) {
//            String description =(element.getAttribute("description").getValue());
//            description = XmlUtil.extractActions(description);
//
//            updateHTMLDescription(element, infoPanel, editor, description);
//        }
        final Message[] messages = XmlUtil.extractAll(new Message[]{new Message(htmlText, Message.MessageType.TEXT_REGULAR)});
        LessonManager lessonManager = LessonManager.getInstance(lesson);
        if (lessonManager != null) lessonManager.addMessage(messages);
    }

    /**
     *
     * @return true if button is updated
     */
    //updateButton(element, elements, lesson, editor, e, document, target, infoPanel);
    protected boolean updateButton(ExecutionList executionList) throws InterruptedException {
        return true;
    }

    protected void initAgainButton(){
    }

    public abstract void execute(ExecutionList executionList) throws Exception;

    protected void startNextCommand(final ExecutionList executionList){
        ApplicationManager.getApplication().invokeLater(() -> {
            try {
                CommandFactory.buildCommand(executionList.getElements().peek()).execute(executionList);
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }
}

File: ide-features-trainer/src/training/commands/Command.kt
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package training.commands

import com.intellij.openapi.application.ApplicationManager
import org.jdom.Element

abstract class Command(val commandType: CommandType) {

  @Suppress("SpellCheckingInspection")
  enum class CommandType {
    TEXT, TRY, ACTION, NOCOMMAND, MOVECARET, TYPETEXT, COPYTEXT, MOUSEBLOCK, MOUSEUNBLOCK, CARETBLOCK, CARETUNBLOCK, SHOWLINENUMBER, WIN, SETSELECTION
  }

  abstract fun execute(executionList: ExecutionList)

  protected fun startNextCommand(executionList: ExecutionList) {
    CommandFactory.buildCommand(executionList.elements.peek(), executionList.documentationMode).execute(executionList)
  }

  protected fun executeWithPollOnEdt(executionList: ExecutionList, function: (Element) -> Unit) {
    val element = executionList.elements.poll()
    ApplicationManager.getApplication().invokeAndWait {
      function(element)
    }
    startNextCommand(executionList)
  }


}
